### PR TITLE
Add a fallback to alternate shards commands

### PR DESCRIPTION
**Pull request by Willamin**
This allows additional functionality to be added to the shards command in a similar way extra commands are added to `git`. Basically, if the shards command isn't found in shards itself, it will run `shards-#{command}`, passing STDIO through. If there is no executable that matches the attempted command, the regular shards usage message is shown.

### CODE REVIEW

**Comment by ysbaddaden** [comment]
Any use cases?

**Comment by Willamin** [comment]
Extensibility. 
Sometimes subcommands don't belong in the core shards project, but would be nice to have and run under the same shards umbrella. This allows the community to create and maintain subcommands without needing approval from the maintainers of shards itself.

For example, rust has implemented a similar scheme in cargo (https://github.com/rust-lang/cargo/wiki/Third-party-cargo-subcommands). Some good subcommand examples from this list include:
  - audit (Audit lockfile for packages with security vulnerabilities)
  - graph (Build GraphViz DOT files of dependency graphs)

Additionally, I would argue that any of these are valid use cases:
  - Add/remove command (https://github.com/crystal-lang/shards/pull/192)
  - Run command (https://github.com/crystal-lang/shards/issues/157)
  - Release command (https://github.com/crystal-lang/shards/issues/98)

**Comment by RX14** [comment]
I'm not all too sure of this. If your tool is separate from shards why not give it a separate command instead of stuffing it under the shards command? I imagine it's just confusing to users what's part of shards and what is not.

**Comment by ysbaddaden** [comment]
Well, `git` provides this and I don't think there are much confusion.

That being said `shards-x` and `shards x` are just one letter, and strictly equivalent —except that the latter may eat up some arguments...

**Comment by Willamin** [comment]
This PR's implementation should allow seamless passthrough of arguments.

**Comment by straight-shoota** [comment]
The benefit of having `shards x` vs `shards-x` is that it can be used to develop features individually and later adopt them into `shards`.

**Comment by ysbaddaden** [comment]
I'll consider this for post v0.8.0 release.

**Comment by Sija** [comment]
We're already post v0.8.0, just a reminder ;)

**Comment by Sija** [comment]
@Willamin Are you willing to continue working on this PR?

**Comment by Willamin** [comment]
@Sija Yes, sorry for the delay, I'll implement the suggestions from @straight-shoota. 😄

File: src/cli.cr
Comment (Willamin) [reply src/cli.cr]: Ah, good call. I like that name better too!

File: src/cli.cr
Comment (Willamin) [reply src/cli.cr]: Excellent suggestion 👍

File: src/cli.cr
Comment (Sija) [inline src/cli.cr:93]: These are redundant

```suggestion
```

Comment (Willamin) [reply src/cli.cr:93]: Given that shards is an important tool, I opted for explicitly using these arguments rather that falling back to the default. This would avoid a problem if the defaults were ever changed. 

That being said, I'm happy to make this change if you think it's a positive change! 😄

Comment (Sija) [reply src/cli.cr:93]: Given that wound be a breaking change I don't see a need for being explicit that much - in the end using the defaults has some merits too (like using `:redirect` pipe instead of explicitly passing `STD*`)

**Comment by RX14** [comment]
I've never actually seen a `git-foo` executable not included in the standard git. Is this feature actually used there?

**Comment by jhass** [comment]
yeah. Two good examples are https://github.com/tummychow/git-absorb and http://hg-git.github.io/. If you want you can also count https://github.com/github/hub, which is supposed to be aliased to `git` and adds new subcommands if you will.

**Comment by straight-shoota** [comment]
And a big collection of extra commands: https://github.com/tj/git-extras

File: src/cli.cr
Comment (Sija) [inline src/cli.cr:86]:
```suggestion
      args: args[1..],
```

**Review by straight-shoota — APPROVED** [review]

**Review by jhass — APPROVED** [review]

File: src/cli.cr
Comment (Sija) [inline src/cli.cr:82]:
```suggestion
          if program_path = Process.find_executable(program_name)
            run_shards_subcommand(program_path, args)
```

**Review by straight-shoota — APPROVED** [review]

**Comment by straight-shoota** [comment]
Thank you @Willamin

**Pull request merged**
